### PR TITLE
Add location picking feature

### PR DESCRIPTION
**Pull request by reloaded**
#1 Add location picking feature

- Added a "Capture Mouse Coordinates" button next to the X, Y coordinate textboxes allowing the user to capture screen coordinates using the mouse. The captured coordinates will be automatically prefilled in the aforementioned textboxes.
- Added a new CaptureMouseScreenCoordinatesWindow that overlays a partially transparent window across all available screens/monitors.
- CaptureMouseScreenCoordinatesWindow exposes a public event that MainWindow hooks in to to receive the user's mouse coordinates upon mouse click.

image

The following scenarios have been tested with no issues.
- 1 monitor.
- 2 monitors, horizontally (see below).
image

- 2 monitors, vertically (see below).
image

- 2 monitors, horizontally, but one has a Y offset (see below).
image

- 2 monitors, vertically, but one has a X offset (see below).
image

The following scenarios have been tested, but with some issues.
- When a multi monitor setup's primary monitor is configured with a negative offset the capture window does not render fully and correctly for some reason. Still looking into this but I'd argue this scenario is not a common one and can be accepted as a buggy scenario which can be fixed further down the road.

The following scenarios have not been tested.
- 3 monitors, where 2 are horizontal and the 3rd is vertically stacked above/under the others.

### CODE REVIEW

**Comment by reloaded** [comment]
@oriash93  bringing your attention to this PR, hopefully it can get merged in relatively soon as it's a helpful feature when wanting to click in a specific spot on screen.

**Comment by reloaded** [comment]
Based on my testing the only bug I found is for the rare scenario where the user's primary screen is configured with a negative offset.

In the horizontal layout configured as seen below (screen 1 is primary).

image

The bug is that the capture window does not draw fully across screen 2. See below.

image

A note about the above image: For some reason the Snipping Tool and Print Screen functions capture things backwards. In the screenshot showing the bug the screen where the window was only rendered halfway across is Screen 2 in the display configuration screenshot...

In a vertical layout configured as seen below (screen 1 is primary).

image

The bug is that the capture window does not draw fully across screen 1. See below.

image


To provide further documentation on the oddity of this bug, I have two identical monitors. Same make/model, same resolution, same physical size. When specifing the primary screen with a negative offset and setting the correct Width/Height and Top/Left values of the mouse capture window, Windows or the .NET framework changes the window size at runtime when the window is rendered. I'm not sure why... See below. I have two screens that are 3840x2160 resolution, as you can see I calculated the correct total width of the window (7680px) yet the rendered window size is 5137.33px.

```
[21:08:45 INF] Opening window to capture mouse coordinates.
[21:08:45 DBG] Total screens detected: 2
[21:08:45 INF] Screen[Bounds={X=0,Y=0,Width=3840,Height=2160} WorkingArea={X=0,Y=0,Width=3840,Height=2160} Primary=True DeviceName=\\.\DISPLAY1
[21:08:45 INF] Screen[Bounds={X=-3840,Y=7,Width=3840,Height=2160} WorkingArea={X=-3840,Y=7,Width=3840,Height=2160} Primary=False DeviceName=\\.\DISPLAY2
[21:08:45 INF] Min Screen X: -3840
[21:08:45 INF] Max Screen X: 0
[21:08:45 INF] Min Screen Y: 0
[21:08:45 INF] Max Screen Y: 7
[21:08:45 INF] Set window size. Width: 7680, Height: 2167
[21:08:45 INF] Setting window position. Left: -3840, Top: 0
[21:08:45 INF] Opened window to capture mouse coordinates.
[21:08:45 INF] Rendered window size: Width: 5137.33333333333, Height: 1462
[21:08:45 INF] Rendered window position: Left:-3840, Height: 1462
[21:08:52 INF] Captured mouse position: 1707, 1836
[21:08:52 INF] Closing window to capture mouse coordinates.\
```

I went into the live tree viewer and manually changed the width to 7680 and nothing happened as far as the window size increasing in width. If I set the window value to something less than 5137.33 than the window width actually reduces in size in real time. Something is capping it and I don't know why.

image

image

image

**Comment by oriash93** [comment]
Hey @reloaded! first of all, thanks a lot for the PR!
I will take a look at it ASAP

File: AutoClicker/AutoClicker.csproj [hunk 1]
```diff
@@ -195,6 +202,7 @@
     <Content Include="Resources\Icons\icon.ico">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </Content>
+    <Resource Include="Resources\Icons\location-crosshairs-solid.ico" />
```
Comment (oriash93) [inline AutoClicker/AutoClicker.csproj]: two questions regarding this:
1. did you create the icon yourself, or did you import it from somewhere?
2. can you please make sure it's defined like the other resources? (configured in the Properties window in VS)

File: AutoClicker/Utils/Constants.cs [hunk 2]
```diff
@@ -1,4 +1,6 @@
-﻿namespace AutoClicker.Utils
+﻿using System.Drawing;
```
Comment (oriash93) [inline AutoClicker/Utils/Constants.cs]: this import is unused here, please remove it

File: AutoClicker/Views/CaptureMouseScreenCoordinatesWindow.xaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+﻿<Window x:Class="AutoClicker.Views.CaptureMouseScreenCoordinatesWindow"
+        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+        xmlns:local="clr-namespace:AutoClicker.Views"
+        mc:Ignorable="d"
+        Title="CaptureMouseScreenCoordinatesWindow" Height="450" Width="800"
```
Comment (oriash93) [inline AutoClicker/Views/CaptureMouseScreenCoordinatesWindow.xaml:8]: should we limit the height and width here?

File: AutoClicker/Views/CaptureMouseScreenCoordinatesWindow.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            base.OnMouseDown(e);
+
+            var position = System.Windows.Forms.Cursor.Position;
+            OnCoordinatesCaptured?.Invoke(this, position);
+            Log.Information($"Captured mouse position: {position.X}, {position.Y}");
+            Close();
+        }
+
+        public event EventHandler<System.Drawing.Point> OnCoordinatesCaptured;
```
Comment (oriash93) [inline AutoClicker/Views/CaptureMouseScreenCoordinatesWindow.xaml.cs]: simplify System.Drawing.Point by adding an import please

**Review by oriash93 — CHANGES_REQUESTED** [review]
Partial review in the meantime, great work so far!

File: AutoClicker/AutoClicker.csproj (hunk 1 above)
Comment (reloaded) [reply AutoClicker/AutoClicker.csproj]: 1. The icon is a FontAwesome free SVG icon converted to ICO format. 
https://fontawesome.com/icons/location-crosshairs?s=solid&f=sharp 

2. Sure.

File: AutoClicker/Views/CaptureMouseScreenCoordinatesWindow.xaml (hunk 3 above)
Comment (reloaded) [reply AutoClicker/Views/CaptureMouseScreenCoordinatesWindow.xaml:8]: The height and width here should not be limited, it's sized dynamically in the code behind. This window/view spans all screens from corner to corner to capture the mouse click location for the auto clicker. Let me know if I misunderstood you here.

**Comment by reloaded** [comment]
@oriash93  I took care of your suggestions & requests. Go ahead and give it another look.

If Github gives you the option to do so, you should ensure this PR is merged in via a Squash merge. That way your main branch only has a single commit that brings in this feature and keeps your commit history clean.

**Comment by reloaded** [comment]
@oriash93  bump, do you have time to finish reviewing this PR?

**Comment by oriash93** [comment]
Hey @reloaded sorry for the delay. I plan to review and merge by the end of this week :)

**Comment by reloaded** [comment]
Great.

**Review by oriash93 — APPROVED** [review]

**Comment by oriash93** [comment]
@reloaded Thank you so much for your contribution!

**Pull request merged**